Supervised revision (#146)
UBC-DSCI/introduction-to-datascience: 06-classification_continued.Rmd

evaluate our classifier without having to visit the hospital to collect more
tumour images? 

> The trick is to split up the data set into a **training set** and **test set**,

Author: for the second one i made a new issue for a global formatting pass

> The trick is to split up the data set into a **training set** and **test set**,

Author: for the first one, done

shown in Figure \@ref(fig:06-training-test)&mdash;and 

> only show the classifier the **training set** when building the classifier.

Author: done

Then to evaluate the accuracy of the classifier, we can use it to predict the
labels (which we know) in the **test set**. If our predictions match the true

> labels for the observations in the **test set** very well, then we have some

Author: done both

the observations in the test set? One way we can do this is to calculate the
**prediction accuracy**. This is the fraction of examples for which the
classifier made the correct prediction. To calculate this we divide the number
of correct predictions by the number of predictions made. 
This process is illustrated in Figure \@ref(fig:06-ML-paradigm-test).
Note that there are other measures for how well classifiers perform, such as *precision* and *recall*; 
these will not be discussed here, but you will likely encounter them in other more advanced
courses on this topic. 

```{r 06-ML-paradigm-test, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Process for splitting the data and finding the prediction accuracy", fig.retina = 2, out.width = "800"}
knitr::include_graphics("img/ML-paradigm-test.png")
```

> In R, we can use the `tidymodels` library collection not only to perform K-nearest neighbour

Reviewer: Throughout the rest of this chapter and the book, can we use "KNN" as opposed to the other acronym flavours? I suggest this as this is how it is referred to in ISLR (https://www.statlearning.com/), the most well know stat/ML learning book in R.

> In R, we can use the `tidymodels` library collection not only to perform K-nearest neighbour

Author: created a new issue #202

classification, but also to assess how well our classification worked. Let's
work through an example of this process using the breast cancer dataset.
We start by loading the necessary libraries, reading in the breast cancer data
from the previous chapter, and making a quick scatter plot visualization of
tumour cell concavity versus smoothness coloured by diagnosis in Figure \@ref(fig:06-precode).

> ```{r 06-precode, fig.height = 4, fig.width = 5, fig.cap="Scatterplot of tumour cell concavity versus smoothness coloured by diagnosis label"}

Author: axis labels i will leave for the detail chapter pass in an issue.

I've removed the cbPalette thing though.

> training data set) and getting an accurate evaluation of its performance (by

Author: no more numbers

> using a larger test data set).  Here, we will use 75% of the data for training,

Author: fixed

and 25% for testing.  To do this we will use the `initial_split` function, 

> specifying that `prop = 0.75` and the response variable is `Class`:

Author: done

```{r 06-initial-split}
set.seed(1)

> cancer_split <- initial_split(cancer, prop = 0.75, strata = Class)

Reviewer: A list here for the steps below (again, GitHub won't let me put inline comments on things that haven't been changed recently).

In 2. Pre-process the data we write that "An additional consideration we need to take when doing this is that we should create the standardization preprocessor using only the training data. This ensures that our test data does not influence any aspect of our model training." but will students remember how data is standaridized? We might need to remind them how this is done, and how including the test data will change/influence this?

Reviewer: Anytime below we write "K-nearest neighbour", let's change it to "KNN".

Reviewer: In 3 should we narrate and remind them that to setup the classifier as before, we need a model spec, and then we combine our recipe from the preprocessing with our model spec in a workflow, which we then pass to `fit` to get our trained classifier object?

Reviewer: At the bottom of 3, in the note we write: 

Note that in your own code, you only have to set the seed once at the beginning of your analysis.

That's not entirely true. In a script yes, in a notebook you did restart + run all yes, but not in a notebook where they might run cells independently multiple times... I think in our assignments we put a `set.seed` before each random function to guard against that...

Reviewer: In the first paragraph from 4, we write "...while the `.pred_class` contains the predicted diagnoses from the model." but I think we should write "...while the `.pred_class` contains the predicted diagnoses from the classifier."

Reviewer: After we call and view the output of `metrics`, I think we should explain that the first row, where it says accuracy, the value for accuracy is in the `.estimate` column. Maybe we should comment on why it is labelled as estimate... For example, why is that a reasonable name (kind of nice foreshadowing to CV that is coming next I think!). Also, we could mention that `kap` is short for Cohen's kappa - another way you can measure classifier performance that they might learn about in more advanced courses.  Also, what does `binary` mean?

Author: rather than tell them about things that we won't be doing anything with, i've just said to focus on the accuracy and ignore the rest (beyond scope).

  metrics(truth = Class, estimate = .pred_class)
```

```{r 06-accuracy-2, echo = FALSE, warning = FALSE}
cancer_acc_1 <- cancer_test_predictions %>% 
                metrics(truth = Class, estimate = .pred_class) %>% 
                filter(.metric == 'accuracy')
```

This shows that the accuracy of the classifier on the test data 
was `r round(100*cancer_acc_1$.estimate, 0)`%.
We can also look at the *confusion matrix* for the classifier, which shows
the table of predicted labels and correct labels, using the `conf_mat` function:

```{r 06-confusionmat}
confusion <- cancer_test_predictions %>%
             conf_mat(truth = Class, estimate = .pred_class)
confusion
```

```{r 06-confusionmat-2, echo = FALSE}
confusionmt <- confusion %>% tidy()
confu11 <- (confusionmt %>% filter(name == "cell_1_1"))$value
confu12 <- (confusionmt %>% filter(name == "cell_1_2"))$value
confu21 <- (confusionmt %>% filter(name == "cell_2_1"))$value
confu22 <- (confusionmt %>% filter(name == "cell_2_2"))$value
```

> This table shows that the classifier labelled 
> `r confu11`+`r confu22` = `r confu11+confu22` observations 
> correctly, `r confu21` observations as benign when they were truly malignant,
> and `r confu12` observations as malignant when they were truly benign. 

Reviewer: Maybe also add a sentence here as why we want to look at this confusion matrix, like why not just look at the overall accuracy? What can this tell us that overall accuracy cannot. UPDATE: I see this below now. Still do the text change I suggest below to make the sentence more digestible to learners.
```suggestion
This table shows that the classifier labelled 
`r confu11`+`r confu22` = `r confu11+confu22` observations 
correctly. It also shows us that it made some mistakes, it labelled `r confu21` observations as benign when they were truly malignant,
and `r confu12` observations as malignant when they were truly benign.  
```

Author: done

                      mutate(percent = 100*n/nrow(cancer_train))
cancer_proportions
```
```{r 06-proportions-2, echo = FALSE, warning = FALSE}
cancer_propn_1 <- cancer_proportions %>%
                filter(Class == 'B') %>%
                select(percent)
```

Since the benign class represents the majority of the training data,
the majority classifier would *always* predict that a new observation
is benign. The estimated accuracy of the majority classifier is usually
fairly close to the majority class proportion in the training data.
In this case, we would suspect that the majority classifier will have 
an accuracy of around `r round(cancer_propn_1[1,1], 0)`%.
The K-nearest neighbours classifier we built does quite a bit better than this, 
with an accuracy of `r round(100*cancer_acc_1$.estimate, 0)`%. 
This means that from the perspective of accuracy,
the K-nearest neighbours classifier improved quite a bit on the basic
majority classifier. Hooray! But we still need to be cautious; in 
this application, it is likely very important not to miss-diagnose any malignant tumours to avoid missing
patients who actually need medical care. The confusion matrix above shows

> that the classifier does indeed miss-diagnose a non-negligible number malignant tumours as benign.

Reviewer: Could we give numbers here, or a percentage to indicate the scale of the mistakes?

Author: done

Therefore, even though the accuracy improved upon the majority classifier,
our critical analysis suggests that this classifier may not have appropriate performance
for the application.

## Tuning the classifier

> 

Author: I want to avoid introducing a bunch of methodology/text/code that ends up not being used by students because "they should do this other thing". I will try to rephrase this bit to make it take less of a leap

which will lead to a better choice of the number of neighbours $K$ for the
overall set of training data. 

> > **Note:** you might be wondering why we can't we use the multiple splits to

Reviewer: I wonder if this note comes too early? I think they at least need to see CV and the CV figure before they might be able to even think about this.

Author: moved it later

> test our final classifier after tuning is done. This is simply because at the
> end of the day, we will produce a single classifier using our overall
> training data. If we do multiple train / test splits, we will end up with
> multiple classifiers, each with their own accuracy evaluated on different
> test data.

Let's investigate this idea in R! In particular, we will use different seed

> values in the `set.seed` function to generate five different train / validation

Reviewer: Our learners have never seen for loops before... This is extra cognitive load this will distract from our point. Our point is that we want to illustrate that different splits will give us different accuracy estimates. I wonder if a simpler way to illustrate this, with less cognitive load, would be to use one seed to calculate one accuracy with code and show it and discuss it in the text. Then in another code chunk, repeat the code (yes it's not DRY code, but that is not the point here) using a different seed and show we get a different accuracy. Then in the text, say if we did this this 3 more times with different seeds (we could even list them in the text) we would get accuracies X, Y and Z (we do the calculation in a hidden code cell). This would avoid the loops, which they don't know, but still get our point across.
If you really want to use a loop here, then I think we have to teach loops in the wrangling chapter.

Also, related to my comment above, I think we should show this under a heading called "Validation accuracy" and then this leads into can we do something better? And yes we can, the better thing is cross validation and we can then explain that. So just moving the CV text above below this.

Author: removed for loop

**cross-validation**.  In **cross-validation**, we split our **overall training
data** into $C$ evenly-sized chunks, and then iteratively use $1$ chunk as the
**validation set** and combine the remaining $C-1$ chunks 
as the **training set**. 

```{r 06-cv-image, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "5-fold cross validation", fig.retina = 2, out.width = "800"}
knitr::include_graphics("img/cv.png")
```

This procedure is shown in Figure \@ref(fig:06-cv-image).
Here, $C=5$ different chunks of the data set are used,
resulting in 5 different choices for the **validation set**; we call this
*5-fold* cross-validation. To do 5-fold cross-validation in R with `tidymodels`, we

> use another function: `vfold_cv`. This function splits our training data into

Reviewer: `strata` argument is used but not explained. We might want to do CV without stratification first? And then have a section on stratification?

Author: clarified

Author: i will make an issue to maybe do stratification as its own section

```{r 06-10-fold}
cancer_vfold <- vfold_cv(cancer_train, v = 10, strata = Class)

vfold_metrics <- workflow() %>%
                  add_recipe(cancer_recipe) %>%
                  add_model(knn_spec) %>%
                  fit_resamples(resamples = cancer_vfold) %>%
                  collect_metrics()
vfold_metrics
```

### Parameter value selection

Using 5- and 10-fold cross-validation, we have estimated that the prediction
accuracy of our classifier is somewhere around `r round(100*(vfold_metrics %>% filter(.metric == "accuracy"))$mean,0)`%. 
Whether that is good or not
depends entirely on the downstream application of the data analysis. In the
present situation, we are trying to predict a tumour diagnosis, with expensive,
damaging chemo/radiation therapy or patient death as potential consequences of
misprediction. Hence, we might like to 
do better than `r round(100*(vfold_metrics %>% filter(.metric == "accuracy"))$mean,0)`% for this application.  

In order to improve our classifier, we have one choice of parameter: the number of

> neighbours, $K$. Since cross-validation helps us evaluate the accuracy of our

Reviewer: Students were really confused about how `grid = 10` chooses the 10 values. It is not well enough explained in the textbook here, and the documentation is NOT accessible:

"If no tuning grid is provided, a semi-random grid (via `dials::grid_latin_hypercube()`) is created with 10 candidate parameter combinations."

So if we don't want to explain this in the text here, we could avoid that by just always giving a data frame of tuning combinations (which we do next in the book and is MUCH easier to explain).

Author: done

> knn_results

Reviewer: The results here show both `accuracy` and `roc_auc`. We should tell them which rows to pay attention to, what `roc_auc` is, and that it is beyond the scope of this book but they might learn about it in more advanced classed. There are some other columns we should probably explain too - or at least tell them to ignore, such as `.estimator`, `n`, `std_error` and `.config`. I think in general, anytime we show them a data frame summary from an analysis, we need to explain it fully, and point out relevant and non-relevant information.

Author: done

  labs(x = "Neighbors", y = "Accuracy Estimate")

> accuracy_vs_k

Reviewer: Can we please make this plot show all of the K's?

Author: done

```
Figure \@ref(fig:06-find-k) suggests that setting the number of 
neighbours to $K =$ `r (accuracies %>% arrange(desc(mean)) %>% head(1))$neighbors`
provides the highest accuracy.
But as you can see, there is no exact or perfect answer here; 
 any selection from $K = 2$ and $15$ would be reasonably justified, as all
of these differ in classifier accuracy by a small amount. Remember: the
values you see on this plot are *estimates* of the true accuracy of our
classifier. Although the 
$K =$ `r (accuracies %>% arrange(desc(mean)) %>% head(1))$neighbors` value is 
higher than the others on this plot,
that doesn't mean the classifier is actually more accurate with this parameter
value! Generally, when selecting $K$ (and other parameters for other predictive
models), we are looking for a value where:

- we get roughly optimal accuracy, so that our model will likely be accurate
- changing the value to a nearby one (e.g. adding or subtracting 1) doesn't decrease accuracy too much, so that our choice is reliable in the presence of uncertainty

> - the cost of training the model is not prohibitive (e.g., in our situation, if $K$ is too large, predicting becomes expensive!)

Reviewer: Can we make it explicitly clear here, that considering those 3 points we would in fact choose K = 12 here? And not K = 3?

Author: done

### Under/Overfitting

To build a bit more intuition, what happens if we keep increasing the number of
neighbours $K$? In fact, the accuracy actually starts to decrease! Rather than
setting `grid = 10` and letting `tidymodels` decide what values of $K$ to try,
let's specify the values explicitly by creating a data frame with a `neighbors`
variable. Figure \@ref(fig:06-lots-of-ks) shows a plot of estimated accuracy as 
we vary $K$ from 1 to almost the number of observations in the data set.

```{r 06-lots-of-ks, fig.height = 4, fig.width = 5, fig.cap="Plot of accuracy estimate versus number of neighbours for many K values"}

> set.seed(1)

Reviewer: In the next line where we do: `k_lots <- tibble(neighbors = seq(from = 1, to = 385, by = 10))` should we show the output of `k_lots`? I am not sure we have shown `seq` before in the book have we? Nor do we often use `tibble`?

Reviewer: Also, the plot below makes the drop off in accuracy look super bad, it's not good, but it's not 0. We should make the y-axis go from 0-1.

Author: seq function is now explained earlier; i disagree with the plotting comment (makes it hard to see what's going on in the plot)

**Strengths:**

1. Simple, intuitive algorithm
2. Few assumptions about what the data must look like 
3. Works for binary (two-class) and multi-class (> 2 classes) classification problems

**Weaknesses:**

1. As the training data gets larger, K-nearest neighbours becomes very slow
2. May not perform well with a large number of predictors
3. Does not perform well when classes are imbalanced (when many more observations are in one of the classes compared to the others)

## Predictor variable selection

> > This section is not required reading for the remainder of the textbook.

Author: agree, edited

Another potentially important part of tuning your classifier is to choose which

> variables from your data will be treated as predictor variables. Based
> on your initial exploratory analysis, you can choose
> anything from using a single predictor variable to using every variable in your
> data; the K-nearest neighbours algorithm accepts any number of

Reviewer: I am not sure what you mean here by "Based on your initial exploratory analysis", as you could (unwisely) put all the variables in your model without doing any exploratory analysis. I see exploratory analysis, in the realm of variable selection, as being useful to rule out any predictors that obviously have little to no predictive value. 

Exploratory analysis is also important for many other things of course (cataloguing missing values, examining class imbalance, feature engineering, etc), but maybe we can just leave it out of this sentence, and instead write:
```suggestion
variables from your data will be treated as predictor variables. 
Technically, you can choose
anything from using a single predictor variable to using every variable in your
data; the K-nearest neighbours algorithm accepts any number of
```

Author: agree

predictors. However, it is **not** the case that using more predictors always
yields better predictions! In fact, including irrelevant predictors can
actually negatively affect classifier performance.

Let's take a look at an example where K-nearest neighbours performs
worse when given more predictors to work with. Here we have modified
the breast cancer data to include the variables `Smoothness` and `Concavity`,

> as well as 500 irrelevant variables

Reviewer: Does it have to be 500 to demonstrate this? That's a really exaggerated number... Also, since KNN suffers from the curse of dimensionality, is this poor prediction due to that, or due to irrelevant predictors?

Author: > curse of dimensionality, is this poor prediction due to that, or due to irrelevant predictors?

the curse of dimensionality in the context of K-NN is precisely the fact that irrelevant, noisy predictors that just add more or less random contributions to the pairwise distances. They're one and the same.

And yeah, we do need 500 (after looking at the plot you'll see why). But I will edit the writing to avoid explicitly stating that. I think it'll make it clearer to avoid mentioning it anyway.

named `Irrelevant1` to `Irrelevant500`, which
we generated ourselves using a random number generator.
You will see from the scatter plot below with two example
variables `Irrelevant1`, `Irrelevant2` that
these irrelevant variables have no meaningful relationship

> with the `Class` response variable.

Reviewer: Here you use response variable, but earlier in the chapter we use target. I think we need to come up with a chosen terminology for this book, and use it consistently. When we first introduce the term, we can qualify the other terms that also mean the same thing. We might want to consider making a dictionary, like we have in MDS: https://ubc-mds.github.io/resources_pages/terminology/ (search for "Equivalence classes").

Author: i agree -- that is part of a good "index" (and why it's really hard to automate index generation). When we finally do formatting stuff and I generate the index, I'll make sure to list all of the versions of these.

I've changed everything to "response" for now, but I have also opened an issue to add "other common words people use" at the point of first introduction as part of the specific chapter revisions.

Author: Issue #165 for this

<center>
```{r 06-irrelevant-loaddata, echo = FALSE, warning = FALSE, fig.cap = "A scatter showing two of the synthetic variables that were added to the data", fig.retina = 2, out.width = "600"}
cancer_irrelevant <- read_csv("data/wdbc_irrelevant.csv") %>%
                mutate(Class = as_factor(Class))

plot_irrelev <- cancer_irrelevant %>%
  ggplot(aes(x = Irrelevant1, y = Irrelevant2, color = Class)) +
  geom_point(alpha = 0.5) +
  labs(color = "Diagnosis") +
  scale_color_manual(labels = c("Malignant", "Benign"), values = cbPalette)

plot_irrelev
```
</center>

Next, we build a sequence of K-NN classifiers that include both `Smoothness` and `Concavity` as predictor variables,
but also increasingly many irrelevant variables. Each model is tuned via 5-fold cross-validation.

> We show the plot of estimated accuracy versus the number of irrelevant predictors below.

Reviewer: ```suggestion
We show the plot of estimated cross-validation accuracy versus the number of irrelevant predictors below.
```

Reviewer: Also, we should reference the figures in the text as we refer to them (instead of saying below).

Author: This is something we should definitely address throughout the whole book -- I will open an issue to remind us to do it

Author: Issue #166

    tune_grid(resamples = cancer_vfold, grid = 10) %>%
    collect_metrics() %>%
    filter(.metric == "accuracy") %>%
    summarize(mx = max(mean)) %>%
    select(mx)
  accs[[i]] <- accs[[i]]$mx
}
accs <- accs %>% unlist()

# get accuracy if we always just guess the most frequent label
base_acc <- cancer_irrelevant %>%
                group_by(Class) %>%
                summarize(n = n()) %>%
                mutate(frac = n/sum(n)) %>%
                summarize(mx = max(frac)) %>%
                select(mx)
base_acc <- base_acc$mx %>% unlist()

# plot
res <- tibble(ks = ks, accs = accs)
res <- res %>% mutate(base_acc = base_acc)
plt_irrelevant_accuracies <- res %>%
  ggplot() +
  geom_line(mapping = aes(x=ks, y=accs, linetype="Tuned K-NN")) +
  geom_hline(data=res, mapping=aes(yintercept=base_acc, linetype="Always Predict Benign (63%)")) +
  labs(x = "Number of Irrelevant Predictors", y = "Model Accuracy Estimate") + 
  scale_linetype_manual(name="Method", values = c("dashed", "solid"))

plt_irrelevant_accuracies
```
</center>

Wow! Once we add sufficiently many irrelevant predictor variables, our classifier
does not do much better than just always guessing "benign", 

>  the most common class label in the data (63% of observations are benign).

Reviewer: Can you expand upon the baseline model ("Always predict Benign") here? How did you calculate this? Were irrelevant predictors added to it? I think this is the first time we have compared a model to this baseline (which I like, and now wonder if we should do it more above...)

Author: @ttimbers what detail are you looking for here? It just always predicts benign regardless of predictors.

You can calculate the accuracy by just computing what fraction of the training data are benign (that's the pct training accuracy you will get).

> Were irrelevant predictors added to it?

Doesn't matter if they were or not, you always predict "benign" no matter what

> which I like, and now wonder if we should do it more above...)

This is probably a really good idea. I think the right strategy would be to introduce it in the 2nd classification chapter, since that focuses on evaluation. I will leave this conversation unresolved for now, because i'll probably try to address it as part of this PR

Author: OK, added a discussion of this earlier on in the evaluation section

> So then if we can't just throw in the kitchen sink, how do we choose which

Reviewer: I like the "kitchen sink" reference, but non-native english speakers might not know this phrasing and it might be confusing? Can you qualify it?

Author: I was thinking the same thing as I wrote it :( we are better off just not using that colloquialism

Author: edited

variables to use as predictors?  This is a very difficult question to answer in
general, and there are a number of methods that have been developed that apply
in particular cases of interest; see the additional resources at the end of

> this chapter to find out where you can learn more about variable selection.

Reviewer: Can we mention that domain expertise is also a way for variable selection? Perhaps here is where we could mention EDA as well, as being another way. However, it is a bit more rudimentary and time consuming.

Author: edited

Here we will discuss *best subset selection,* which is applicable to any 
classification method (K-NN or otherwise), and works when you have a small

> number of total variables from which you want to keep only the relevant ones (to

Reviewer: What is a small number? Your example above had 502 predictors... Is that small? Would best subset work there? Maybe the explanation is just around the corner, but as someone who has never read this before I am wondering now.

Author: defined small as "at most 10" immediately

3. finally pick the subset of predictors that gives you the highest estimated prediction accuracy.  

Note that best subset selection becomes very slow when you have even a moderate
number of predictors to choose from. This is because the number of possible predictor subsets
grows very quickly with the number of predictors. For example, if we have $m=2$ predictors---let's call
them A and B---then we have 3 variable sets to try: A alone, B alone, and finally A
and B together. If we have $m=3$ predictors---A, B, and C---then we have 7
to try: A, B, C, AB, BC, AC, and ABC. In general, the number of models
we have to train for $m$ predictors is $2^m-1$; in other words, when we 
get to $m=10$ predictors we have over *one thousand* models to train, and 
at $m=20$ predictors we have
over *one million* models to train! If you have more than about 10 predictors,
you should first use your scientific understanding of the data
to choose the 10 or so predictors that you think are most meaningful, and then
run best subset selection within those 10. Alternatively, you can use a different
predictor variable selection method; see the additional resources at the end

> of this chapter. 

Reviewer: OK, I see here you define what small vs large are. So best subset will not work for your 502 example above likely. 

Also, what about number of observations... that will impact this as well, but not in the same way. Should we mention that?

Author: edited

(I removed explicit reference of the 500 variables here)

We now turn to implementing best subset selection in R.
Unfortunately there is no built-in way to do this using the `tidymodels` framework,

> so we will have to code it ourselves. First we will use the `select` function

Reviewer: This is a lot of code... I would do this myself (not thinking of this course) in a similar way, but use nested data frames, `map`, and write a function to get the accuracy - more in the `tidyverse` style of doing things. The nested data frames and `map` are something we do use in the clustering chapter, but what we are doing here is more complicated... 

I will code up my `tidyverse` style of this and paste it below to compare. But I think both solutions are more complicated than what we want in this book, and the idea we discussed earlier of abstracting this to an R package is probably the right move.

Reviewer: Here is how I would roughly write this in the tidyverse style:

```
library(tidyverse)
library(tidymodels)

get_model_accuracy <- function(model_formula_predictors) {
  # create a model specification
  knn_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) %>%
    set_engine("kknn") %>%
    set_mode("classification")
  # create a 5-fold cross-validation object
  cancer_vfold <- vfold_cv(cancer_subset, v = 5, strata = Class)
  # create a recipe from the model string
  cancer_recipe <- recipe(as.formula(paste("Class ~", model_formula_predictors)), data = cancer_subset) %>%
    step_scale(all_predictors()) %>%
    step_center(all_predictors())
  # tune the K-NN classifier with these predictors, and collect the accuracy for the best K
  acc <- workflow() %>%
    add_recipe(cancer_recipe) %>%
    add_model(knn_spec) %>%
    tune_grid(resamples = cancer_vfold, grid = 10) %>%
    collect_metrics() %>%
    filter(.metric == "accuracy") %>%
    summarize(mx = max(mean))
  acc$mx %>% unlist()
}

cancer <- read_csv("data/wdbc.csv")
cancer_train <- cancer
cancer_subset <- cancer_train %>% select(Class, Smoothness, Concavity, Symmetry, Area)

possible_predictors = colnames(cancer_subset %>% select(-Class))

best_subset <- tibble(n_pred = 2:4, 
                      poss_pred = list(possible_predictors)) %>% 
  mutate(combinations = map2(poss_pred, n_pred, combn, simplify=FALSE)) %>% 
  unnest(combinations) %>% 
  mutate(formula_predictors = map(combinations, ~paste(., collapse="+"))) %>% 
  unnest(formula_predictors) %>% 
  mutate(accuracy = map_dbl(formula_predictors, get_model_accuracy))
```
I might be able to simplify it a bit more, but not much. I think it's too complicated for the book, so I still vote to abstract it to a package.

Reviewer: I can simplify it to this with the new `dplyr` `rowwise` function:

```
library(tidyverse)
library(tidymodels)

get_model_accuracy <- function(model_formula_predictors) {
  # create a model specification
  knn_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) %>%
    set_engine("kknn") %>%
    set_mode("classification")
  # create a 5-fold cross-validation object
  cancer_vfold <- vfold_cv(cancer_subset, v = 5, strata = Class)
  # create a recipe from the model string
  cancer_recipe <- recipe(as.formula(paste("Class ~", model_formula_predictors)), data = cancer_subset) %>%
    step_scale(all_predictors()) %>%
    step_center(all_predictors())
  # tune the K-NN classifier with these predictors, and collect the accuracy for the best K
  acc <- workflow() %>%
    add_recipe(cancer_recipe) %>%
    add_model(knn_spec) %>%
    tune_grid(resamples = cancer_vfold, grid = 10) %>%
    collect_metrics() %>%
    filter(.metric == "accuracy") %>%
    summarize(mx = max(mean))
  acc$mx %>% unlist()
}

cancer <- read_csv("data/wdbc.csv")
cancer_split <- initial_split(cancer_train, prop = 0.75, strata = Class)
cancer_subset <- training(cancer_split) %>% 
  select(Class, Smoothness, Concavity, Symmetry, Area)

possible_predictors = colnames(cancer_subset %>% select(-Class))

best_subset <- tibble(n_pred = 2:4) %>% 
  rowwise() %>% 
  mutate(combo = list(combn(possible_predictors, n_pred, simplify=FALSE))) %>% 
  unnest(combo) %>% 
  rowwise() %>% 
  mutate(formula_predictors = paste(combo, collapse="+"),
         accuracy = get_model_accuracy(formula_predictors))
```

Author: @ttimbers the chapter now uses forward selection -- any idea how to simplify that?

Author: closing this thread since we've moved to forward selection and @ttimbers is going to spin that out into a separate package.

        # create a recipe from the model string
        cancer_recipe <- recipe(as.formula(model_string), data = cancer_subset) %>%
            step_scale(all_predictors()) %>%
            step_center(all_predictors())

  
        # tune the K-NN classifier with these predictors, and collect the accuracy for the best K
        acc <- workflow() %>%
          add_recipe(cancer_recipe) %>%
          add_model(knn_spec) %>%
          tune_grid(resamples = cancer_vfold, grid = 10) %>%
          collect_metrics() %>%
          filter(.metric == "accuracy") %>%
          summarize(mx = max(mean))
        acc <- acc$mx %>% unlist()

        # add this result to the dataframe
        accuracies <- accuracies %>% add_row(size = i, index = j, model_string = model_string, accuracy = acc)
    }
}

accuracies
```
The maximum accuracy estimate is obtained for the formula
`Class ~ Concavity + Symmetry + Area`, so the best subset selection 
method would result in a model with these three predictors. Note,
however, that the model with all 4 predictors, as well as the 
model `Class ~ Smoothness + Concavity + Area` are reasonable

> alternatives that provide a similar estimated accuracy.

Reviewer: This result - that we should include all predictors doesn't bring home the message we started this chapter out with... I think we should see if there is an irrelevant predictor in that data set (although, from what I know of it, I don't think there is) or add in 1-2 of the irrelevant predictors and show that best subset selection will in fact remove them. Right now as the chapter is, we haven't demonstrated that, just told them that it will...

Author: fixed now!

completely different.  In general, if the model *is influenced too much* by the

> training data, it is said to **overfit** the data.

Reviewer: Should we mention that such a classifier will not generalize well to new data?

Author: says that in the previous sentence

You can see these two effects in Figure \@ref(fig:06-decision-grid-K),
which shows how the classifier changes as we set the number of neighbours $K$ to 1, 7, 20, and 200.

<center>

> ```{r 06-decision-grid-K, echo = FALSE, fig.height = 7, fig.width = 10, fig.cap = "Effect of K in overfitting and underfitting"}

Reviewer: Can we show an even more extreme K, say 300? Where it gets as bad as it can? I think that might help get the point across better.

Author: done

> randomly into a training and test data set, using the training set to build the

Reviewer: The splitting data section above seems out of place where it is... Should it go above where we split into train and test?

Reviewer: Also, from the docs it seems that `initial_split` randomly splits by default (and if you don't want to do that, as in a time series, you would use `initial_time_split`)... 

UPDATE: I now read that we are saying this, but it's not clear to me in the first read through. Can we make it more explicit in this sentence: "The tidymodels function initial_split and vfold_cv functions do this for us."

Reviewer: I think the explanation for stratification should be improved, we do not explicitly say how using the strata argument ensures that enough of each class ends up in each of the train, validation, and test partitions.

Reviewer: ## Summary section:

"Classification algorithms use one or more quantitative variables to predict the value of a third, categorical variable. " shouldn't say "third" because we say "one or more", it should say "another"

Reviewer: "To tune the classifier (e.g., select the K in K-nearest neighbours), we maximize accuracy estimates from cross-validation. " should say "We can tune the classifier (e.g., select the K in K-nearest neighbours) by maximizing accuracy estimates from cross-validation. " I suggest this change as we could use some other metric to maximize, we just choose to use accuracy here for simplicity in this book.

Author: done

classifier, and using the test set to estimate its accuracy. To tune the
classifier (e.g., select the K in K-nearest neighbours), we maximize accuracy
estimates from cross-validation. This process is summarized in Figure \@ref(fig:06-overview).

```{r 06-overview, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Overview of K-nn classification", fig.retina = 2, out.width = "660"}
knitr::include_graphics("img/train-test-overview.jpeg")
```

> The overall workflow for performing K-nearest neighbour classification using `tidymodels` is as follows:

Reviewer: These numbered steps do not correspond to the other 6 numbered steps above... Maybe we shouldn't number the ones above?

Author: done

1. Use the `initial_split` function to split the data into a training and test set. Set the `strata` argument to the response variable. Put the test set aside for now.

> 2. Use the `vfold_cv` function to split up the training data for cross validation.

Reviewer: I think we should swap steps 2 & 3 here? Based on the order we introduce them in the book?

Author: i think it's fine as is

3. Create a `recipe` that specifies the response and predictor variables, as well as preprocessing steps for all variables. Pass the training data as the `data` argument of the recipe.
4. Create a `nearest_neighbors` model specification, with `neighbors = tune()`.
5. Add the recipe and model specification to a `workflow()`, and use the `tune_grid` function on the train/validation splits to estimate the classifier accuracy for a range of $K$ values.

> 6. Pick a value of $K$ that yields a high accuracy estimate that doesn't change much if you change $K$ to a nearby value.

Reviewer: Change 7 to say "Make a new model specification for the best parameter value (i.e. $K$) and retrain the classifier using the `fit` function."

Author: done

**Strengths:**

> 

Reviewer: I feel like we need a bit of a introductory sentence here, something like "All algorithms have strengths and weakenesses...."

Author: done

1. is a simple, intuitive algorithm
2. requires few assumptions about what the data must look like 
3. works for binary (two-class) and multi-class (more than 2 classes) classification problems

**Weaknesses:**

The K-nearest neighbours classifier:

1. becomes very slow as the training data gets larger
2. may not perform well with a large number of predictors
3. may not perform well when classes are imbalanced 

## Predictor variable selection

> *This section is not required reading for the remainder of the textbook. It is included for those readers 
interested in learning how irrelevant variables can influence the performance of a classifier, and how to
pick a subset of useful variables to include as predictors.*

Another potentially important part of tuning your classifier is to choose which
variables from your data will be treated as predictor variables. Technically, you can choose
anything from using a single predictor variable to using every variable in your
data; the K-nearest neighbours algorithm accepts any number of
predictors. However, it is **not** the case that using more predictors always
yields better predictions! In fact, including irrelevant predictors can

> actually negatively affect classifier performance.

Author: agree

**The effect of irrelevant predictors**

Let's take a look at an example where K-nearest neighbours performs
worse when given more predictors to work with. Here we have modified
the breast cancer data to have only the `Smoothness`, `Concavity`, and `Perimeter` variables from the original data,
and then added irrelevant variables that we created ourselves using a random number generator.
The irrelevant variables each take a value of 0 or 1 with equal probability for each observation, regardless
of what the value `Class` variable takes. In other words, the irrelevant variables have 

> no meaningful relationship with the `Class` variable.

Author: edited

```{r 06-irrelevant-gendata, echo = FALSE, warning = FALSE}
set.seed(4)
cancer_irrelevant <- cancer %>% select(Class, Smoothness, Concavity, Perimeter)

for (i in 1:500) {
    # create column
    col = (sample(2, size=nrow(cancer_irrelevant), replace=TRUE)-1)
    cancer_irrelevant <- cancer_irrelevant %>% 
	add_column( !!paste("Irrelevant", i, sep="") := col)
}
```
```{r 06-irrelevant-printdata, warning = FALSE}
cancer_irrelevant %>% 
      select(Class, Smoothness, Concavity, Perimeter, Irrelevant1, Irrelevant2)
```

> Next, we build a sequence of K-NN classifiers that include `Smoothness`,

Author: done

`Concavity`, and `Perimeter` as predictor variables, but also increasingly many irrelevant
variables (from 0 to 40 of them). Each model is tuned via 5-fold
cross-validation. Figure \@ref(fig:06-performance-irrelevant-features) shows
the estimated cross-validation accuracy versus the number of irrelevant predictors.  As
we add more irrelevant predictor variables, the estimated accuracy of our
classifier decreases. This is because the irrelevant variables add a random
amount to the distance between each pair of observations; the more irrelevant
variables there are, the more (random) influence they have, and the more they
corrupt the set of nearest neighbours that vote on the class of the new
observation to predict.  

<center>

> ```{r 06-performance-irrelevant-features, echo = FALSE, warning = FALSE, fig.retina = 2, out.width = "600", fig.cap = "Effect of inclusion of irrelevant predictors"}

Author: i disagree, it makes it much harder to see what's going on in the plot.

still outperforms the baseline majority classifier (with about `r round(cancer_propn_1[1,1], 0)`% accuracy) 
even with 40 irrelevant variables.
How could that be? Figure \@ref(fig:06-neighbours-irrelevant-features) provides the answer:
the tuning procedure for the K-nearest neighbours classifier combats the extra randomness from the irrelevant variables 
by increasing the number of neighbours. Of course, because of all the extra noise in the data from the irrelevant
variables, the number of neighbours does not increase smoothly; but the general trend is increasing.
Figure \@ref(fig:06-fixed-irrelevant-features) corroborates
this evidence; if we fix the number of neighbours to $K=3$, the accuracy falls off more quickly.

```{r 06-neighbours-irrelevant-features, echo = FALSE, warning = FALSE, fig.retina = 2, out.width = "600", fig.cap = "Tuned number of neighbours for varying number of irrelevant predictors"}
plt_irrelevant_nghbrs <- ggplot(res) +
              geom_line(mapping = aes(x=ks, y=nghbrs)) +
              labs(x = "Number of Irrelevant Predictors", y = "Number of neighbours")
plt_irrelevant_nghbrs
```

> ```{r 06-fixed-irrelevant-features, echo = FALSE, warning = FALSE, fig.retina = 2, out.width = "600", fig.cap = "Accuracy versus number of irrelevant predictors for tuned and untuned number of neighbours"}

Author: not worth it

res_tmp <- res %>% pivot_longer(cols=c("accs", "fixedaccs"), names_to="Type", values_to="accuracy")
plt_irrelevant_nghbrs <- ggplot(res_tmp) +
              geom_line(mapping = aes(x=ks, y=accuracy, color=Type)) +
              labs(x = "Number of Irrelevant Predictors", y = "Accuracy") + 
              scale_color_discrete(labels= c("Tuned K", "K = 3"))
plt_irrelevant_nghbrs
```

> 

Reviewer: Are we getting into too many details in this section? Not sure, but just want to give pause and think and answer this question. The main point we want to get across is that extra predictors can sometimes negatively affect your predictions for the reasons you state above. The issue about outperforming the baseline classifier, while interesting  to us and true, might add too much complexity for our audience? We might want to sacrifice a little truth here for clarity? We could just not mentions the performance of the baseline classifier, and then figs 7.10 and 7.11, and the details you describe about them. 

Again, not sure we should do that, but we should at least think it to be sure we want to get this detailed.

Author: this is already in an optional section -- if the reader is here they're already digging in. i'm leaving it as is

**Finding a good subset of predictors**

> So then if we can't just use all of our variables as predictors without consideration, how 

Author: agree

do we choose which variables we *should* use?  A simple method is to rely on your scientific understanding
of the data to tell you which variables are not likely to be useful predictors. For example, in the cancer

> data that we have been studying, the `ID` variable is just a unique identifier for the observation.

Author: not certain it's for a patient -- it's for a tumour image. But I feel like (i.e. tumour image) is more confusing than just not having anything there.

As it is not related to any measured property of the cells, the `ID` variable should therefore not be used
as a predictor. That is, of course, a very clear-cut case. But the decision for the remaining variables 
is less obvious, as all seem like reasonable candidates. It 
is not clear which subset of them will create the best classifier. One could use visualizations and
other exploratory analyses to try to help understand which variables are potentially relevant, but
this process is both time-consuming and error-prone when there are many variables to consider.
We therefore need a more systematic and programmatic way of choosing variables. 
This is a very difficult problem to solve in
general, and there are a number of methods that have been developed that apply
in particular cases of interest; see the additional resources at the end of

> this chapter to find out where you can learn more about variable selection.

Author: revised

> The first idea you might think of for a systematic way to select predictors
> is *best subset selection*. Best subset selection essentially involves trying
> all possible subsets of variables and picking the "best" one. In particular, you

Reviewer: ```suggestion
The first idea you might think of for a systematic way to select predictors
is trying all possible subsets of variables and picking the "best" one.
This is a bona fide variable selection method which is commonly referred to 
as *best subset selection*. To perform best subset selection, you:
```

Author: done

1. create a separate model for every possible subset of predictors
2. tune each one using cross validation

> 3. pick the subset of predictors that gives you the highest estimated prediction accuracy  

Reviewer: should we make this clear that this is cross-validated prediction accuracy?

Author: done

Best subset selection is applicable to any  classification method (K-NN or otherwise).
However, it becomes very slow when you have even a moderate
number of predictors to choose from (say, around 10). This is because the number of possible predictor subsets
grows very quickly with the number of predictors, and you have to train the model (itself
a slow process!) for each one. For example, if we have $m=2$ predictors---let's call
them A and B---then we have 3 variable sets to try: A alone, B alone, and finally A

> and B together. If we have $m=3$ predictors---A, B, and C---then we have 7

Reviewer: Do we need  "m ="  here? It might be adding unnecessary cognitive overload for new learners here.

Reviewer: Update - I see you do 2^m-1 and m^2 later, so maybe it's needed?

Author: I removed it everywhere except the formula

to try: A, B, C, AB, BC, AC, and ABC. In general, the number of models
we have to train for $m$ predictors is $2^m-1$; in other words, when we 
get to $m=10$ predictors we have over *one thousand* models to train, and 
at $m=20$ predictors we have over *one million* models to train! 
So although it is a simple method, best subset selection is usually too computationally 
expensive to use in practice.

Another idea is to iteratively build up a model by adding one predictor variable 

> at a time. This method&mdash;known as *forward selection*&mdash;is also widely

Reviewer: ISLR calls this forward stepwise selection, so let's do the same to be consistent?
```suggestion
at a time. This method&mdash;known as *forward stepwise selection*&mdash;is also widely
```

Author: I've seen it much more commonly referred to as "forward selection"

applicable and fairly straightforward. It involves the following steps:

1. start with no predictors
2. run the following 4 steps until you run out of predictors:
    1. for each predictor you have not chosen yet, try adding it to the set of predictors to make a candidate model
    2. tune each candidate model
    3. pick the model with the highest estimated accuracy
    4. add the new predictor for that model to your set

> 3. pick the set of predictors with a good trade-off between accuracy and model simplicity

Reviewer: I think we need to work on the clarity of this... Learners who don't yet know this algorithm might not know exactly what we mean by "predictor not yet chosen", and "predictor set".... Here's how they explain it in ISLR, we can maybe use this as a base, but turn the math into words?

<img width="556" alt="Screen Shot 2021-07-13 at 3 19 13 PM" src="https://user-images.githubusercontent.com/6865016/125532548-5a547a65-bc1c-4780-8a80-8daf146b5c3f.png">

Author: if you turn the math in ISLR into words, you end up with essentially what I have. It's quite difficult to keep track of the temporary models / etc that forward selection creates using just words. I've tried to edit it, but I'm not sure I made it any clearer.

Author: i'll open an issue to revisit it if it is not sufficiently clear

Say you have $m$ total predictors to work with. In the first iteration, you have to make
$m$ candidate models, each with 1 predictor. Then in the second iteration, you have
to make $m-1$ candidate models, each with 2 predictors (the one you chose before and a new one).
This pattern continues for as many iterations as you want. If you run the method
all the way until you run out of predictors to choose, you will end up training

> approximately $m^2$ separate models. This is a *big* improvement from the $2^m-1$

Reviewer: Should we give some numbers here like we do above and compare? Say p =10? With best subset we said we'd need to train over one thousand models, here with forward stepwise we'd only have to train 100.

Author: done

As an example, let's make a model formula for all the predictors,
which should output something like
`Class ~ Smoothness + Concavity + Perimeter + Irrelevant1 + Irrelevant2 + Irrelevant3`:
```{r}
example_formula <- paste("Class", "~", paste(names, collapse="+"))
example_formula
```

Finally, we need to write some code that performs the task of sequentially
finding the best predictor to add to the model.
If you recall the end of the wrangling chapter, we mentioned
that sometimes one needs more flexible forms of iteration than what 
we have used earlier, and in these cases one typically resorts to
[a for loop](https://r4ds.had.co.nz/iteration.html#iteration).
This is one of those cases! Here we will use two for loops:
one over increasing predictor set sizes 
(where you see `for (i in 1:length(names))` below),
and another to check which predictor to add in each round (where you see `for (j in 1:length(names))` below).
For each set of predictors to try, we construct a model formula,
pass it into a `recipe`, build a `workflow` that tunes
a K-NN classifier using 5-fold cross-validation, 
and finally records the estimated accuracy.

> ```{r 06-fwdsel-2, warning = FALSE}

Reviewer: We definitely need to tidy this, or abstract to a function in a package, I am thinking of a function with this spec:

```
add_variable_selection <- function(workflow, method = "forward", metric = "accuracy") {
  # return tibble with model objects, model formula and accuracies
}
```

What do you think?

Author: @ttimbers sounds good

I'm not going to wait until this happens to merge the PR. You can make a separate PR for that when you're done.

        cancer_recipe <- recipe(as.formula(model_string), data = cancer_subset) %>%
                          step_scale(all_predictors()) %>%
                          step_center(all_predictors())

        # tune the K-NN classifier with these predictors, and collect the accuracy for the best K
        acc <- workflow() %>%
          add_recipe(cancer_recipe) %>%
          add_model(knn_spec) %>%
          tune_grid(resamples = cancer_vfold, grid = 10) %>%
          collect_metrics() %>%
          filter(.metric == "accuracy") %>%
          summarize(mx = max(mean))
        acc <- acc$mx %>% unlist()

        # add this result to the dataframe
        accs[[j]] <- acc
        models[[j]] <- model_string
    }
    jstar <- which.max(unlist(accs))
    accuracies <- accuracies %>% add_row(size = i, model_string = models[[jstar]], accuracy = accs[[jstar]])
    selected <- c(selected, names[[jstar]])
    names <- names[-jstar]
}
accuracies
```

Interesting! The forward selection procedure first added the three meaningful variables `Perimeter`,
`Concavity`, and `Smoothness`, followed by the irrelevant variables. Figure \@ref(fig:06-fwdsel-3)

> visualizes the accuracy as a function of the number of predictors in the model. You can see that

Reviewer: Not sure our learners will follow "... as a function of ..."
```suggestion
visualizes the accuracy versus the number of predictors in the model. You can see that
```

as meaningful predictors are added, the estimated accuracy increases substantially; and as you add irrelevant
variables, the accuracy either exhibits small fluctuations or decreases as the model attempts to tune the number
of neighbours to account for the extra noise. In order to pick the right model from the sequence, you have 

> to trade off the desiderata to have a high accuracy, but also a simple model with few predictors (which has 

Reviewer: desiderata is not a commonly used word and provides unnecessary cognitive load in this sentence.
```suggestion
to balance high accuracy and model simplicity - one which has a few predictors (which has 
```

Author: done

> a lower chance of overfitting). The way to perform that trade-off is to look for the *elbow*

Reviewer: Is it an elbow? It's in the wrong orientation for me to call it that, I'd say the peak? Or something that reflects we are looking for the top of an upside down u-shape.

Author: I prefer elbow -- we aren't looking for the peak (that's the maximum). The "elbow" is where there is a first "kink" in the plot (it's described in the text as where the plot starts to "level off").

in Figure \@ref(fig:06-fwdsel-3), i.e., the place on the plot where the accuracy stops increasing dramatically and
levels off or begins to decrease. The elbow in Figure \@ref(fig:06-fwdsel-3) appears to occur at the model with 
3 predictors; after that point the accuracy levels off. So here the right trade-off of accuracy and number of predictors
occurs with 3 variables: `Class ~ Perimeter + Concavity + Smoothness`. In other words, we have successfully removed irrelevant
predictors from the model! It is always worth remembering, however, that what cross validation gives you 
is an *estimate* of the true accuracy; you have to use your judgement when looking at this plot to decide
where the elbow occurs, and whether adding a variable provides a meaningful increase in accuracy.

> Remember: since the choice of which variables to include as predictors is
> part of tuning your classifier, you *cannot use your test data* for this
> process! 

> ```{r 06-fwdsel-3, warning = FALSE, fig.retina = 2, out.width = "600", fig.cap = "Estimated accuracy versus the number of predictors for the sequence of models built using forward selection"}

Reviewer: I don't think we need to show this code chunk to the learners.

Author: done
